Cognitive Services swagger api-version 2017-04-18 initial commit

felixwa opened the pull request:
1. Support Cognitive Services RP version - 2017-04-18
2. Add following operations and related parameter definitions:
a. POST updateAccountsCreationSettings
b. GET accountsCreationSettings
c. POST checkSkuAvailability
d. Get operations
3. Fix all issues reported by AutoRest/Validate-Spec/Validate-Example.
4. Examples are removed from this swagger file – will add them to
dedicated example folder later.

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

felixwa committed:
version 2017-04-18 initial commit

1. Support Cognitive Service RP version - 2017-04-18
2. Add following operations and related parameter definitions:
a. POST updateAccountsCreationSettings
b. GET accountsCreationSettings
c. POST checkSkuAvailability
d. Get operations
3. Fix all issues reported by AutoRest/Validate-Spec/Validate-Example.
4. Examples are removed from this swagger file – will add them to
dedicated example folder later.

xiaomin-ms commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+            "required": true,
+            "x-ms-client-flatten": true,
+            "schema": {
+              "$ref": "#/definitions/CognitiveServicesResourceKindSettingUpdateParameter"
```
ResourceKindSetting => AccountsCreationSettings?

felixwa replied:
Changed.

xiaomin-ms commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+        "allowCreate": {
+          "type": "string",
+          "description": "Whether allow user to create resource.",
+          "enum": [
```
Does it support Boolean type?

felixwa replied:
It supports but AutoRest suggests to use Enum instead of Boolean.

veronicagg replied:
should this be a boolean? could this property accept other values?

veronicagg replied:
the suggestion from AutoRest is to evaluate if the property is really a boolean or not, the intent is to consider if there could be more than 2 values possible for the property in the future or not, if the answer is no, then a boolean is fine, if the answer is yes, there could be other values, then make it an enum with values that represent what those should be. The idea is that it can help avoid breaking changes if you want to add one more possible value, naming the values "true", "false" I believe defeats this purpose.

felixwa replied:
Changed to Boolean.

felixwa replied:
I suggest AutoRest can turn off this warning though.

salameer commented:
Hi @felixwa 

@alvadb  will be reviewing this PR very shortly, thanks for your patience! :)

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+            "pattern": "^[a-zA-Z0-9][a-zA-Z0-9_.-]*$"
+          },
+          {
+            "name": "body",
```
should this be named "parameters"?

felixwa replied:
Changed all places to "parameters".

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+      "post": {
+        "tags": [ "CognitiveServicesAccounts" ],
+        "operationId": "CheckSkuAvailability_List",
+        "description": "Check availalable SKUs.",
```
typo in "available"

felixwa replied:
Fix.

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+          "200": {
+            "description": "OK.",
+            "schema": {
+              "$ref": "#/definitions/CheckSkuAvailabilityResultList"
```
is this a pageable operation?

felixwa replied:
No.

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+    "/subscriptions/{subscriptionId}/providers/Microsoft.CognitiveServices/locations/{location}/updateAccountsCreationSettings": {
+      "post": {
+        "tags": [ "CognitiveServicesAccounts" ],
+        "operationId": "CognitiveServices_UpdateResourceKindSettings",
```
is this operation id accurate for the operation path?

felixwa replied:
Fixed.

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+            "$ref": "#/parameters/locationParameter"
+          },
+          {
+            "name": "body",
```
should this be named "parameters"?

felixwa replied:
Changed all places to "parameters".

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "OK"
```
is there no model returned from this operation?

felixwa replied:
Right, no model returned for this operation.

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+      "get": {
+        "tags": [ "CognitiveServicesAccounts" ],
+        "operationId": "AccountsCreationSettings_List",
+        "description": "Read resource kind settings.",
```
is this description accurate for this operation path?

felixwa replied:
Fixed.

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+            "required": true,
+            "type": "string",
+            "enum": [ "current" ],
+            "description": "Accounts creation setting name, now only support \"current\".",
```
are you planning on adding other values? If there's only one value in the enum on a required parameter, then it's value will be taken as a constant (https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#single-value-enum-as-a-constant) 
You could just write that value in the path as a constant, if you're not planning on adding more values.

felixwa replied:
This operation has been changed to POST after more discussions with ARM team. This "current" has gone.

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/CognitiveServicesResourceKindSettingResult"
```
is this a pageable operation?

felixwa replied:
No.

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1071 @@
+        "skuAvailable": {
+          "type": "string",
+          "description": "Indicates the given SKU is available or not.",
+          "enum": [
```
should this be a boolean? are there other values this property may accept?

felixwa replied:
Changed to Boolean.

veronicagg commented:
@felixwa thanks for submitting this PR. 
If this spec is based on a previous api-version:
"When you are creating the swagger spec for the new api-version, please copy the swagger spec from the previous version in to the new api-versioned folder and commit it. After that overwrite it with the changes for the new api-version. This makes it easy for us to review the changes."
(https://github.com/Azure/azure-rest-api-specs/blob/c12accd4df379643a3aa9c5bf385e6f6ef50bf73/.github/CONTRIBUTING.md#filenames-and-folder-structure) 
It's great to see linter validation is clean. Please consider if you'll be adding descriptions for the remaining warnings, as those would help with documentation (https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096052#L1022). 
For a new api-version, we're requiring x-ms-examples, to make sure validation is clean against those.

veronicagg reviewed (changes requested):
Comments inline.

felixwa committed:
Commit 2016-02-01-preview version for comparison

Commit original 2016-02-01-preview version for comparison purpose

felixwa committed:
Fix review comments

Fix review comments from veronicagg

felixwa committed:
Fix a typo

Fix a typo.

felixwa commented:
@veronicagg , I have commit 2016-02-01-preview version then made another commit to overwrite it with 2017-04-18 version. Let me know if this works for you.
Also I have added those missing "description" fields - it is strange that AutoRest I use locally did not report them, I'm sure I am using the latest. Does travis-ci use a different version?
As for ms-examples, I know now they should be added to a separated folder. Please help review existing changes first and if they looking good for you I can add them in this PR or a new PR (I prefer to add them in a new PR).

veronicagg commented:
@felixwa thanks for committing the previous version! Regarding the latest version of autorest, here are the update instructions for autorest: https://github.com/Azure/autorest#updating-autorest (just to make sure you're running the latest - Travis should be using the latest). 
The examples (x-ms-examples) are in a separate folder, but need a reference from the swagger spec in each operation. We should do it in this PR, because it helps catch validate the swagger spec, so if there are issues, we want to fix them from the start instead of bug fixes (as this is a new api-version). Thanks!

veronicagg commented:
@felixwa should you be adding "securityDefinitions" section, similar to https://github.com/Azure/azure-rest-api-specs/blob/master/arm-analysisservices/2016-05-16/swagger/analysisservices.json#L26 ?

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1055 @@
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/CognitiveServicesAccountCreateParameters"
```
>CognitiveServicesAccountCreateParameters [](start = 37, length = 40)

Is there a reason for the body parameter and the response not being the same model? Please consider this rule" "M2017: The model definition for the body parameter and the response MUST be the same for a PUT operation. This ensures that the same entity will be reusable between GET and PUT." (from [https://github.com/Azure/azure-rest-api-specs/blob/62c96876795e600de8369efe42edd3a70f494017/documentation/swagger-checklist.md](https://github.com/Azure/azure-rest-api-specs/blob/62c96876795e600de8369efe42edd3a70f494017/documentation/swagger-checklist.md))

felixwa replied:
@veronicagg , we followed RPC v2. In PUT body, there is no need to put id/name/type etc. there because they are already there in URL.

veronicagg replied:
@felixwa would you consider changing this in the future? Users of the api won't be able to reuse the same identity between GET and PUT ("This ensures that the same entity will be reusable between GET and PUT." ). There will be an error shown by our validation tools on this soon.

felixwa replied:
Ok we could consider that if the tool will report this anyway. Will that be a breaking change for RP and/or this swagger file?

veronicagg replied:
It would be a breaking change for the SDK, so it'd need to be done consciously.

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1055 @@
+        }
+      },
+      "required": [
+        "name"
```
Please take a look at: "M2016: For PATCH operation, every property in the model for the request body MUST be optional and the property MUST NOT provide any default value. The absence of default values applies recursively to complex properties in the model definition." from [https://github.com/Azure/azure-rest-api-specs/blob/62c96876795e600de8369efe42edd3a70f494017/documentation/swagger-checklist.md](https://github.com/Azure/azure-rest-api-specs/blob/62c96876795e600de8369efe42edd3a70f494017/documentation/swagger-checklist.md)

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1055 @@
+      "properties": {
+        "sku": {
+          "$ref": "#/definitions/Sku",
+          "description": "Required. Gets or sets the SKU of the resource."
```
Please take a look at: "M2016: For PATCH operation, every property in the model for the request body MUST be optional and the property MUST NOT provide any default value. The absence of default values applies recursively to complex properties in the model definition." (from [https://github.com/Azure/azure-rest-api-specs/blob/62c96876795e600de8369efe42edd3a70f494017/documentation/swagger-checklist.md](https://github.com/Azure/azure-rest-api-specs/blob/62c96876795e600de8369efe42edd3a70f494017/documentation/swagger-checklist.md))

felixwa replied:
Not required. I have revised the "description".

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1055 @@
+      ],
+      "description": "The SKU of the cognitive services account."
+    },
+    "CognitiveServicesAccountPropertiesCreateParameters": {
```
>CognitiveServicesAccountPropertiesCreateParameters [](start = 5, length = 50)

is there no type? for this model? object?

felixwa replied:
This is just a placeholder actually. The RP SDK requires this field but we don't have any properties in it.

veronicagg replied:
I'm not sure I understand this, could you clarify? "properties" is also a required property, what would the user input to this property? what would you include in your example for this case?

felixwa replied:
Like below:
--------------
{
	"sku":{
		"name": "S0"
	},
	"location": "westeurope",
	"kind": "Face",
	"properties": {}
}
--------------------
User would input nothing in the "properties" field. For our RP it is not required, it is just RP SDK requires that field present there.

veronicagg commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1055 @@
+        },
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/CognitiveServicesAccountPropertiesCreateParameters",
```
>CognitiveServicesAccountPropertiesCreateParameters [](start = 33, length = 50)

what is this flattening? I don't see any properties in the referenced model.

felixwa replied:
Removed (but AutoRest reports a new warning).

veronicagg commented:
:clock1:

veronicagg reviewed (changes requested):
Please see some more questions/comments inline.

felixwa committed:
Fix some issues and review comments

1. Change account name max/min length to be consistent with UI.
2. Resolved some review comments.
3. Fix some issues with CheckSkuAvailability operation.

felixwa commented:
@veronicagg , I have added securityDefinitions and resolved some review comments.
I can add examples in this PR after we agree on left open ones.

veronicagg reviewed (changes requested):
Please see reply to a couple of your comments.

felixwa committed:
Adding examples

Adding examples for every operation.
Fix some issues found when adding examples.

felixwa commented:
@veronicagg , I have added examples for every operation.

felixwa commented on arm-cognitiveservices/2017-04-18/swagger/cognitiveservices.json:
```diff
@@ -0,0 +1,1137 @@
+        "message": {
+          "type": "string",
+          "description": "Additional error message."
+        }
```
@veronicagg , I noticed that validate-example reports warnings for these 2 fields because the in example they are null. I tried to make the type ["string", "null"] but then AutoRest reports errors then.
Please advice how to make tools happy in this case.

veronicagg replied:
@felixwa you may be hitting https://github.com/Azure/azure-rest-api-specs/issues/798, if so, please ignore the reported error for now.

veronicagg commented:
Regarding the "properties" that I asked before, is there a reason why there is a $ref in that case and not something like?
        "properties": {
          "type": "object",
          "additionalProperties": {
            "type": "string"
          },
          "description": "Must exist in the request. Must not be null."
        }

or even without "additionalProperties"?
Also, you may to indicate in the description that an empty object must be sent in this case.

felixwa committed:
Remove CognitiveServicesAccountPropertiesCreateParameters

Remove CognitiveServicesAccountPropertiesCreateParameters, make it
"inline" and refine description.
We still need it because currently the underlying SDK requires it.

felixwa commented:
@veronicagg , I have removed that CognitiveServicesAccountPropertiesCreateParameters and make it "inline" as you suggested.

AutorestCI commented:
No modification for Ruby

AutorestCI commented:
No modification for NodeJS

AutorestCI commented:
No modification for Python
